### DISCUSSION TITLE
Generation of APIs (e.g. swagger/openAPI)

### DISCUSSION THREAD

**Discussion by cmungall** [linkml/linkml#241] [Q&A] [answered]
I would like to autogenerate a swagger/openAPI CRUD interface from a LinkML schema.

This would implement a number of patterns:

 - CRUD methods for every class
      - exclude by default: abstract classes, mixins
      - exclude? anything without an id or compound key: https://github.com/linkml/linkml-model/pull/16
      - follow standard REST patterns, or more explicit methods?
          - e.g. PUT for C/U
          - for GET: `/person/{id}` -> returns Person class
 - Search methods for any string
      - potentially introduce a specific basic type for "lexical string"
      - e.g. /person/search/name/{searchterm}`

The idea is that something like https://github.com/biolink/biolink-api could be generated directly from the biolink schema

Note that in future it may be possible to generate the actual access methods e.g. for a sql database, mongo database, triplestore. But to scope this, we would start with just the interface.

One possibility is to auto-generate flask code for stub methods for each API call, and then use existing tools to generate swagger.json from this.

Note: we also have a GraphQL exporter that could be used an alternative

**Comment by sierra-moxon** [comment, marked as answer]
https://github.com/linkml/linkml-dataops now supports a generic API framework.
